## Chp 4 字符串与正则表达式

4.1 字符串
* 记住一些转义字符：\0 \a \b \t \n \v \f \r \e
* 字符串属于不可变序列类型：支持序列通用方法，包括切片

In [1]:
年龄 = 19
print(年龄**2)
print(id(年龄))

361
4368213784


* 字符串驻留机制
  * 将其赋值给多个不同的对象时，内存中只有一个副本，多个对象共享该副本。
  * 保护.pyc文件不会被错误代码搞的过大，例如有人写了'abc'\*10\*\*10这种代码
  * 数值-5~256也遵循驻留机制

In [14]:
# 字符串驻留机制 string interning
s = "good"
print(id(s))
t = "good"
print(id(t))

a = -6
b = -6
print(id(a) == id(b))

a = 257
b = 257
print(id(a) == id(b))

4421293296
4421293296
False
False


* 判断变量s是否为字符串：isinstance(s, str)

In [15]:
# 判断变量s是否为字符串
a = 'good'
print(isinstance(a,str))
b = '中国梦'
print(isinstance(b,str))

True
True


* 字符串属于不可变序列，可以进行切片操作

In [16]:
# 字符串属于不可变序列
# 下面进行对字符串的切片操作
s = 'Fudan University'
x = s[2]
y = s[3:8] # 始终记住，切片操作包含起始索引，不包括结束索引
z = s[-1:-6:-1] # 逆序需要确定负的步长
print(x)
print(y)
print(z)

a = 'abc'
c = a + 'xyz'
print(c)
print('x' in c)
print('xy' in c)
print(len(c))

d
an Un
ytisr
abcxyz
True
True
6


* 练习：把输入的字符串切割成指定行数

In [23]:
s = input("请输入一串字符：")
m = eval(input("请输入行数："))

n = len(s) // m if len(s) % m == 0 else len(s) // m +1

for i in range(0, len(s), n):
    print(s[i:i+n])

请输入一串字符：Fudan University
请输入行数：3
Fudan 
Univer
sity


* 字符串常量

In [29]:
import string
print(string.digits)
print(string.punctuation)
print(string.ascii_letters)
print(string.printable)
print(string.ascii_lowercase)
print(string.ascii_uppercase)

0123456789
!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~
abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ
0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!"#$%&'()*+,-./:;<=>?@[\]^_`{|}~ 	

abcdefghijklmnopqrstuvwxyz
ABCDEFGHIJKLMNOPQRSTUVWXYZ


* 练习：删除字符串中的非字符、非数字符号

In [32]:
import string
s = input('请输入字符串：')

character = string.ascii_letters + string.digits

t = ''

for i in s: 
    if i in character: # 要用in，不能用==
        t = t + i
print(t)

请输入字符串：a.b`c-d~1^2<3?4
abcd1234


* 字符串转换函数

In [34]:
s = 'What is your name?'

print(s.upper())
print(s.lower())
print(s.capitalize())
print(s.swapcase()) # 交换大小写
print(s)

WHAT IS YOUR NAME?
what is your name?
What is your name?
wHAT IS YOUR NAME?
What is your name?


* 互逆的两个函数：ord()和chr()，ACSII码和字符串相互转换

In [39]:
print(ord('8')) # 字符串转ASCII
print(chr(56)) # ASCII转字符串

56
8


* 练习：自己实现大小写转换

In [42]:
def MySwap(x): 
    y = x
    if 'a' <= x <= 'z': 
        y = chr(ord(x) - ord('a') + ord('A')) # 这里x不需要打引号，本身x就是作为一个变量传进来，如何打上引号，ord('x')则表示'x'的ASCII值
    if 'A' <= x <= 'Z':  
        y = chr(ord(x) - ord('A') + ord('a'))
    return y

a = input('请输入字符串：')
t = ''

for i in a: 
    t = t + MySwap(i)
print(t)

请输入字符串：Fudan
fUDAN


* 字符串链接函数sep.join(a)

In [43]:
a = ['a', 'b', 'c', 'd']
sep = ','
s = sep.join(a)
print(s)

a,b,c,d


In [48]:
[s, t] = ['abc', 'xyz']
print(''.join([s, t]))
print('-'.join([s, t, s, t]))
print(s.join(t)) # 注意！
print(t.join(s)) # 注意！

abcxyz
abc-xyz-abc-xyz
xabcyabcz
axyzbxyzc


* 练习：输入一个字符串，用空格隔开每一个字符

In [50]:
print(' '.join(input('请输入字符串：')))

请输入字符串：Fudan University
F u d a n   U n i v e r s i t y


* 内置函数eval()：获取<b style="color: red;">字符串表达式</b>值

In [54]:
import math
print(eval('math.sqrt(3)')) # 【字符串表达式】的值！

1.7320508075688772


* str()：将数字转化为字符串

In [62]:
print(str(3.4))
print(str(2.3e5))
print(2.3E-5)

3.4
230000.0
2.3e-05


* 练习：列出1000以内的所有回文数

In [63]:
result = []
for i in range(1000): 
    s = str(i)
    if s == s[::-1]: 
        result.append(s)
print(result)

['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '11', '22', '33', '44', '55', '66', '77', '88', '99', '101', '111', '121', '131', '141', '151', '161', '171', '181', '191', '202', '212', '222', '232', '242', '252', '262', '272', '282', '292', '303', '313', '323', '333', '343', '353', '363', '373', '383', '393', '404', '414', '424', '434', '444', '454', '464', '474', '484', '494', '505', '515', '525', '535', '545', '555', '565', '575', '585', '595', '606', '616', '626', '636', '646', '656', '666', '676', '686', '696', '707', '717', '727', '737', '747', '757', '767', '777', '787', '797', '808', '818', '828', '838', '848', '858', '868', '878', '888', '898', '909', '919', '929', '939', '949', '959', '969', '979', '989', '999']


* <b style="color: red;">加密和解密：非常重要！！！</b>
  * 已知密文和m，如何恢复明文？

In [112]:
# 加密
password = list(input('请输入原文：'))
key = eval(input('请输入钥匙m：'))

answer = []

index = 0

while password: 
    index = (index + key - 1)%len(password)
    answer.append(password.pop(index))
answer = ''.join(answer)
print(answer)

# 解密
z = list(range(len(answer)))
raw = [None]*len(answer)

for i in answer: 
    index = (index + m - 1)%len(z)
    raw[z.pop(index)] = i
raw = ''.join(raw)
print(raw)

请输入原文：abcde
请输入钥匙m：3
caebd
abcde


* 练习（报数出圈）：设有编号为1~n的n个人按顺时针站成一个圆圈。首先从第1个人开始，按顺时针从1开始报数，报到第m个人，令其出列。然后再从出列的下一个人开始，按顺时针从1开始报数，报到第m个人，再令其出列，……。如此下去，直到圆圈不再有人为止。求这n个人出列的顺序。
  * 思路：假设此圈共有rest个人，对应x\[0\], x\[1\], x\[2\], …, x\[rest-1\]。假设这一轮报1的人对应的元素下标为 curIndex (0 <= curIndex < rest)，则报m的人对应的数组元素下标为 (curIndex + m - 1) % rest 。把报m的人对应的元素从列表中删除(元素删除导致 rest 自动减1)。把curIndex赋值为刚才被删除元素的下标，之后进行下一轮报数。

In [77]:
n = eval(input('请输入总人数：'))
m = eval(input('请输入m值：'))

x = list(range(1, n+1))
y = []

currentIndex = 0

while x: 
    currentIndex = (currentIndex + m - 1)%len(x)
    y.append(x.pop(currentIndex)) # x pop掉，append到y后
print(y)

请输入总人数：20
请输入m值：7
[7, 14, 1, 9, 17, 5, 15, 4, 16, 8, 20, 13, 11, 10, 12, 19, 6, 18, 2, 3]


In [98]:
# 明文加密
x = input('请输入一串字符：')
x = list(x)
m = eval(input('请输入密码m：'))
y = []

curIndex = 0
while x: 
    curIndex = (curIndex + m - 1)%len(x)
    y.append(x.pop(curIndex))

y = ''.join(y)
print(y)

请输入一串字符：Fudan
请输入密码m：3
dFnua


In [104]:
# 密文解密
x = input('请输入一串密文：')
x = list(x)
m = eval(input('请输入密码m：'))
y = [None]*len(x)
z = list(range(len(y))) # 指示原来被pop出的位置

curIndex = 0
for i in x: 
    curIndex = (curIndex + m - 1) % len(z)
    y[z.pop(curIndex)] = i

y = ''.join(y)
print(y)

请输入一串密文：dFnua
请输入密码m：3
Fudan


* 字符串对齐方法

In [116]:
print('Fudan'.center(20))
print('Fudan'.center(20, '='))
print('Fudan'.ljust(20, '='))
print('Fudan'.rjust(20, '='))

       Fudan        
=======Fudan========
Fudan===============
===============Fudan


* 练习：打印三角形（1）

In [118]:
lines = eval(input('请输入n值：'))

for i in range(lines): 
    print('*'*(i+1))
print()

请输入n值：3
*
**
***



* 练习：打印三角形（2）

In [119]:
lines = eval(input('请输入n值：'))

for i in range(lines): 
    stars = '*'*(i+1)
    stars = ' '.join(stars)
    print(stars)
print()

请输入n值：3
*
* *
* * *



* 练习：打印三角形（3）

In [122]:
# 方法1
lines = eval(input('请输入n值：'))

for i in range(lines): 
    stars = ' '*(lines - (i+1)) + '*'*(i+1)
    stars = ' '.join(stars)
    print(stars)
print()

请输入n值：3
    *
  * *
* * *



In [129]:
# 方法2
lines = eval(input('请输入n值：'))

for i in range(lines): 
    stars = '*'*(i+1)
    stars = ' '.join(stars)
    stars = stars.rjust(lines*2-1, ' ')
    print(stars)
print()

请输入n值：3
    *
  * *
* * *



* 练习：打印三角形（4）

In [135]:
lines = eval(input('请输入行数：'))

for i in range(lines // 2): 
    stars = '*'*(i + 1)
    stars = ' '.join(stars)
    oneline = stars.rjust(lines, ' ')
    print(oneline)
    
for i in range(lines // 2 - 1, 0, -1): 
    stars = '*'*i
    stars = ' '.join(stars)
    oneline = stars.rjust(lines, ' ')
    print(oneline)

请输入行数：9
        *
      * *
    * * *
  * * * *
    * * *
      * *
        *


* 练习：打印三角形（5）

In [138]:
lines = eval(input('请输入行数：'))

for i in range(lines//2): 
    stars = '*'*(i+1)
    stars = ' '.join(stars)
    oneline = stars.center(lines, ' ')
    print(oneline)

for i in range(lines//2 - 1, 0, -1): 
    stars = '*'*i
    stars = ' '.join(stars)
    oneline = stars.center(lines, ' ')
    print(oneline)

请输入行数：9
    *    
   * *   
  * * *  
 * * * * 
  * * *  
   * *   
    *    


* 字符串格式化
  * Chp 4 ppt p36
  * %s：字符串(采用str()的显示)
  * %r：字符串(采用repr()的显示)
  * %c：单个字符
  * %b：二进制整数
  * %d：十进制整数
  * %i：十进制整数
  * %o：八进制整数
  * %x：十六进制整数
  * %e：指数(基底写为e)
  * %E：指数(基底写为E)
  * %f、%F：浮点数
  * %g：指数(e)或浮点数(根据显示长度)
  * %G：指数(E)或浮点数(根据显示长度)
  * %%：字符“%”

In [146]:
x = 1234
so = '%o' %x # 转化为八进制
print(so)

di = '%d' %x # 十进制转化为十进制
print(di)

X = '%x' %x
print(X)

e = '%e' %x
print(e)

st = '%s' %x
print(st)

2322
1234
4d2
1.234000e+03
1234


* <b style="color: red;">常用格式举例</b>
  * 10.2f：格式化浮点数，总宽度为10，四舍五入到小数点后第2位
  * 10.2e：以科学计数法表示浮点数，总宽度为10，系数保留到小数点后2位
  * 5d：将整数格式化为总宽度为5的十进制形式
  * 5o：将整数格式化为总宽度为5的八进制形式
  * 5x：将整数格式化为总宽度为5的十六进制形式
  * 5b：将整数格式化为总宽度为5的二进制形式
  * 50s：将字符串格式化为总宽度为50的字符串，不足则补上空格
  * 50.2s：截取字符串的前2位，将其格式化为总宽度为50的字符串，不足补空格
  * -10.2f：向左对齐格式化对象，若无负号(即默认)则是右对齐<b style="color: red;">【左边的数为总长度，包括小数点，右边的数为小数点后保留的位数】</b>
  * +10.2f：格式化浮点数，如果是正数，则自动加上正号+（默认不加正号）

* <b style="color: red;">字符串格式化：format()</b>

In [148]:
print( "The number {0:,} in hex is: {0:#x}, the number {1} in oct is {1:#o}".format( 5555, 55 ))
print( "The number {1:,} in hex is: {1:#x}, the number {0} in oct is {0:#o}".format( 5555, 55 ))
print( "my name is {name}, my age is {age}, and my QQ is {qq}".format(name = "Dong Fuguo",age = 37,qq = "306467355"))

The number 5,555 in hex is: 0x15b3, the number 55 in oct is 0o67
The number 55 in hex is: 0x37, the number 5555 in oct is 0o12663
my name is Dong Fuguo, my age is 37, and my QQ is 306467355


In [149]:
position = (5,8,13)
print( "X:{0[0]};Y:{0[1]};Z:{0[2]}".format(position))

X:5;Y:8;Z:13


* <b style="color: red;">方法对象</b>

In [155]:
weather = [("Monday","rain"),("Tuesday","sunny"),("Wednesday", "sunny"),("Thursday","rain"),("Friday","Cloudy")]
formatter = "Weather of '{0[0]}' is '{0[1]}'".format  #方法对象
for item in map(formatter,weather): 
    print(item)

print()

for item in weather: 
    print(formatter(item))

Weather of 'Monday' is 'rain'
Weather of 'Tuesday' is 'sunny'
Weather of 'Wednesday' is 'sunny'
Weather of 'Thursday' is 'rain'
Weather of 'Friday' is 'Cloudy'

Weather of 'Monday' is 'rain'
Weather of 'Tuesday' is 'sunny'
Weather of 'Wednesday' is 'sunny'
Weather of 'Thursday' is 'rain'
Weather of 'Friday' is 'Cloudy'


* <b style="color: red;">%和.format()的区别</b>

In [181]:
x, y = 3.5, 142.26732335

patStrA = '数%8.2f与数%-8.3f之和为%08.2f!' # -是左对齐，默认右对齐
patStrB = '数{0:8.2f}与数{1:<8.3f}之和为{2:08.2f}!' # .format()中左右对其为<和>，默认右对齐

s = patStrA % (x, y, x*y)
t = patStrB.format(x, y, x*y)

print('s=', s)
print('t=', t)

s= 数    3.50与数142.267 之和为00497.94!
t= 数    3.50与数142.267 之和为00497.94!


* 字符串的查找
  * find()、rfind()
    * find()和rfind方法分别用来查找一个字符串在另一个字符串指定范围（默认是整个字符串）中首次和最后一次出现的位置，如果不存在则返回-1；
  * index()、rindex()
    * index()和rindex()方法用来返回一个字符串在另一个字符串指定范围中首次和最后一次出现的位置，如果不存在则抛出异常；
  * count() 
    * count()方法用来返回一个字符串在另一个字符串中出现的次数。

In [187]:
s="apple,peach,banana,peach,pear"
print(len(s))
print(s.find("peach"))
print(s.find("peach",7)) # 在(7, 29)中找
print(s.find("peach",7,20)) # 不存在，返回-1
print(s.rfind('p'))
print(s.index('p'))
print(s.index('pe'))
print(s.index('pear'))
# print(s.index('ppp'))
print(s.count('p'))
print(s.count('pp'))
print(s.count('ppp'))

29
6
19
-1
25
1
6
25
5
1
0


* 字符串分割
  * split()、rsplit()
    * split()和rsplit()方法分别用来以指定字符为分隔符，将字符串左端和右端开始将其分割成多个字符串，并返回包含分割结果的列表；split()与join()作用相反。split()与rsplit()的区别，在第2个参数指定的分隔符个数n小于实际分隔符数目时，才能看出来。
  * partition()、rpartition()
    * partition()和rpartition()用来以指定字符串为分隔符将原字符串分割为3部分元组，即分隔符前的字符串、分隔符字符串、分隔符后的字符串，如果指定的分隔符不在原字符串中，则返回原字符串和两个空字符串。

In [201]:
s="apple,peach,banana,pear"
print(s.split(","))

print(s.partition(',')) # 返回三组字符串

print(s.rpartition(',')) # 返回三组字符串

print(s.rpartition('banana'))

s = "2014-10-31"
t=s.split("-")
print(t)

print(list(map(int, t)))

['apple', 'peach', 'banana', 'pear']
('apple', ',', 'peach,banana,pear')
('apple,peach,banana', ',', 'pear')
('apple,peach,', 'banana', ',pear')
['2014', '10', '31']
[2014, 10, 31]


* 对于split()和rsplit()方法，如果不指定分隔符，则字符串中的任何空白符号（包括空格、换行符、制表符等等）都将被认为是分隔符，返回包含最终分割结果的列表。
  * <b style="color: red;">顺序为找到第一个非空白字符，再开始分割，遇到第一个非空白字符，完成分割</b>

In [204]:
s = 'hello world \n\n My name is Dong   '
print(s.split())

s = '\n\nhello world \n\n\n My name is Dong   '
print(s.split())

s = '\n\nhello\t\t world \n\n\n My name\t is Dong   '
print(s.split())

['hello', 'world', 'My', 'name', 'is', 'Dong']
['hello', 'world', 'My', 'name', 'is', 'Dong']
['hello', 'world', 'My', 'name', 'is', 'Dong']


* split()和rsplit()方法还允许指定最大分割次数
  * <b style="color: red;">重申：顺序为找到第一个非空白字符，再开始分割，遇到第一个非空白字符，完成分割</b>

In [208]:
s = '\n\nhello\t\t world \n\n\n My name is Dong   '
print(s.split(None,1))     # None表示不指定分割符
# 这里直接忽略了开头的\n

print(s.rsplit(None,1))

print(s.split(None,2))

print(s.rsplit(None,2))

print(s.split(None,5))

print(s.split(None,6))

['hello', 'world \n\n\n My name is Dong   ']
['\n\nhello\t\t world \n\n\n My name is', 'Dong']
['hello', 'world', 'My name is Dong   ']
['\n\nhello\t\t world \n\n\n My name', 'is', 'Dong']
['hello', 'world', 'My', 'name', 'is', 'Dong   ']
['hello', 'world', 'My', 'name', 'is', 'Dong']


* 字符串替换

In [210]:
# 查找替换 replace(.)
s="中国，中国"
print(s)
s2 = s.replace("中国", "中华人民共和国")
print(s2)

中国，中国
中华人民共和国，中华人民共和国


* 字符串转换
  * 生成映射表函数maketrans(.)和按映射表关系转换字符串函数translate(.)

In [221]:
import string
#table = str.maketrans('abcdef123','uvwxyz@#$')
s="Python is a greate programming language. I like it!"
print(s.translate(table))

#print(s.translate(table,"gtm")) #第二个参数表示要删除的字符
# python 3中代码发生了一些变化

Python is u gryuty progrumming lunguugy. I liky it!


* 字符串消减
  * strip()、rstrip()、lstrip()：这几个方法分别用来删除两端、右端或左端的空格或连续的指定字符

In [222]:
s=" abc  "
s2=s.strip( )
print(s2)

print("aaaassddf".strip("a"))

print("aaaassddf".strip("af"))

print("aaaassddfaaa".rstrip("a"))

print("aaaassddfaaa".lstrip("a"))

abc
ssddf
ssdd
aaaassddf
ssddfaaa


* 字符成员判断
  * s.startwith(t)、s.endswith(t)：判断字符串是否以指定字符串开始或结束

In [244]:
print("a" in "abcde")

print("j" in "abcde"）

SyntaxError: invalid character '）' (U+FF09) (2579249403.py, line 3)

In [225]:
import os
# [filename for filename in os.listdir(r'c:\\') if filename.endswith(('.bmp','.jpg','.gif'))]

* 检验字符串是否为字母、数字等
  * isalnum()：检验字符串是否为数字或字母
  * isalpha()：检验字符串是否为字母
  * isdigit()：检验字符串是否为数字字符
  * isspace()：检验字符串是否为空白字符
  * isupper()：检验字符串是否为大写字母
  * islower()：检验字符串是否为小写字母

In [233]:
print('1234abcd'.isalnum())
print('1234abcd'.isalpha())
print('1234'.isdigit())
print('1234.0'.isdigit()) # 注意！
print('abcd.'.isalpha())

True
False
True
False
False


* 练习：连续输入若干行字符串（以空行结束），输出最长的那些行。

In [239]:
strList = []

s = input()

while s: 
    strList.append(s)
    s = input()
    
lenList = [len(x) for x in strList]
maxlen = max(lenList)
for i in range(len(strList)): 
    if lenList[i] == maxlen: 
        print(strList[i])

123
asdfgh
zxcvbn
asd
iop

asdfgh
zxcvbn


* 练习：统计文章的单词个数。约定单词由英文字母组成，其他字符只是用来分隔单词。
  * <b style="color: red;">记住这种条件语句的逻辑！</b>

In [245]:
import string
word = 0
inword = False
s = input()

for c in s: 
    if inword: # 避免重复计数，两个条件语句都是碰到另一种情况才会跳转到对方语句
        if c not in string.ascii_letters: 
            inword = False
    else: 
        if c in string.ascii_letters: 
            inword = True
            word += 1
print(word)

abc, cdf, 12a.bc
4


<b style="color: red;">格式化输出期末考试大概率会有</b>

### 4.2 正则表达式
* 记住元字符：
. \d \D \s[\t\n\r\f\v] \S \w[a-zA-Z0-9_] \W ^ \\$ \b \B ^ $ . \* \? \+ \- \ \{ \} \[ \] | ( )  
  * 如果要作为普通字符使用，需要转义
* re模块提供了正则表达式操作所需要的功能：
  * re.findall(pattern, string)：以列表形式列出字符串中模式的所有匹配项

#### 4.2.1 正则表达式元字符：字符类
* 字符类：由一对[]方括号括起来的字符集合，定义方式如下：
  * \[xyz\]:枚举字符集，匹配括号中任意字符
  * '\[pjc\]ython'可以匹配'python'、'jython'、'cython'
  * \[^xyz\]:否定枚举字符集，匹配不在括号中任意字符
  * '\[^abc\]'可以匹配一个任意除'a'、'b'、'c'之外的字符
  * \[a-z\]:指定范围的字符，匹配指定范围的任意字符
  * '\[a-zA-Z0-9\]'可以匹配一个任意大小写字母或数字
  * \[^m-z\]:指定范围以外的字符，匹配指定范围以外的任意字符
* 预定义字符类：正则表达式将常常用到的一些特定字符类形成了若干预定义字符类
  * .：匹配除换行符以外的任意单个字符
  * \d：匹配任何数字，相当于\[0-9\]
  * \D：与\d含义相反,非数字，相当于\[^0-9\]
  * \s：匹配任何空白字符，相当于\[\t\n\r\f\v\]
  * \S：与\s含义相反，相当于\[^\t\n\r\f\v\]
  * \w：匹配任何字母、数字以及下划线，相当于\[a-zA-Z0-9_\]
  * \W：与\w含义相反,相当于\[^a-zA-Z0-9_\]
* 边界匹配符：字符串匹配往往涉及从某个位置开始匹配，例如行的开头或结果、单词边界等，边界匹配符用于匹配字符串的位置
  * ^：匹配行首，匹配以^后面的字符开头的字符串，如：“^a”匹配“abc”中的“a”,不匹配“bat”中的“a”
  * \\$：匹配行尾，匹配以\\$之前的字符结束的字符串，如：“c\\$”匹配“abc”中的“c”,不匹配“acb”中的“c”
  * \b：匹配单词头或单词尾，如：“\\\bfoo\\\b”匹配”foo”、“foo.” “(foo)” “bar foo baz”但不匹配“foobar”“foo3”。注意：’\\b’在正则表达式表示单词边界，而在字符串中’\b’表示退格字符，所以这些与标准转义字符重复的元字符必须使用两个\\字符，在python中也可以使用原始字符串r””或r’’，即“\\\bfoo\\\b”  或 r“\\bfoo\\b”
  * \B：与\b含义相反，如：’py\B’ 匹配 ”python” “py3” “py2”,但不匹配”happy” ”sleepy” ”py!”
* 重复限定符：指定重复的次数（默认最大匹配）
  * X\{n,m\}：X重复n到m次，如：“o\{1,3\}”匹配“fooooood”中的前3个“o”和后3个“o”
  * X\{n,\}：X至少重复n次，如：“o\{2,\}”匹配“fooooood”中的所有的“o”,不匹配“bob”中的“o”
  * X\{n\}，X重复n次，如：“\\b\[0-9\]\{3\}”匹配”000” ~ “999”, “o{2}”匹配“food”中的两个”o”,不匹配“bob”中的“o”
  * X+：X重复1次或多次，等价于X \{1,\}，如：’zo+’ 匹配 ”zo” ,“zoo”,但不匹配 ”z”
  * X\*：X重复0次或多次，等价于X\{0,\}，如：’zo*’ 匹配 ”zo” ,“zoo”, ”z”
  * X?：X重复0次或1次，等价于X\{0,1\}，如：’colou?r’ 匹配 ”color”  ,”colour”；

In [7]:
import re
print(re.findall('o{1,3}', 'fooooood'))

['ooo', 'ooo']


\\bfoo\\b或r'\bfoo\b'，r''为原始字符串

* ‘\*’ 要求前面的字符出现大于等于0次
* '+' 要求前面的字符出现大于等于1次
* '?' 要求前面的字符出现0次或1次
* \[0-9\]\{2,3\} 匹配最少2次最多3次的0-9的数字
* \[0-9\]\{2,\} 匹配至少2次的0-9的数字

In [5]:
import re
re.findall('red|green|blue', 'pink red, green and blue')

['red', 'green', 'blue']

* 匹配算法：贪婪性匹配算法
  * Python针对重复限定符，默认采用贪婪性匹配算法
  * 贪婪性匹配算法是指重复限定符会导致正则表达式引擎尽可能多地重复前导字符，只有当这种重复引起整个正则表达式匹配失败的情况下，引擎会进行回溯

In [6]:
import re
re.findall('<.+>','<book><title>Python</title><author>Dong</author></book>')

['<book><title>Python</title><author>Dong</author></book>']

In [7]:
import re
re.findall('<.+?>','<book><title>Python</title><author>Dong</author></book>')

['<book>', '<title>', '</title>', '<author>', '</author>', '</book>']

* 分组符“()”：重复限定符重复前导字符，如果需要重复的符合某种模式的多个字符，则需要将描述该模式的正则表达式放在()内，()表示一个分组（子模式），即()内的内容作为一个整体出现
  * re. findall(pattern,string\[,flags\]):返回匹配结果列表，若pattern含有子模式，同时返回子模式的列表

In [248]:
telNumber = '''Suppose my Phone No. is 0535-1234567, yours is 010-12345678, his is 025-87654321.'''
pattern = re.compile(r'(\d{3,4})-(\d{7,8})')
print(pattern.findall(telNumber))

print(re.findall(r'(\d{3,4})-(\d{7,8})',telNumber))

print(re.findall(r'((\d{3,4})-(\d{7,8}))',telNumber))

[('0535', '1234567'), ('010', '12345678'), ('025', '87654321')]
[('0535', '1234567'), ('010', '12345678'), ('025', '87654321')]
[('0535-1234567', '0535', '1234567'), ('010-12345678', '010', '12345678'), ('025-87654321', '025', '87654321')]


In [1]:
# compile的源代码
def compile(pattern, flags=0):
    "Compile a regular expression pattern, returning a Pattern object."
    return _compile(pattern, flags)

* 选择符
  * 选择符”|”，用于选择匹配多个可能的正则表达式中的一个
  * 选择符”|”的优先级最低，如果需要使用()来限制选择符的作用范围

In [2]:
import re
print(re.findall('(red|blue|green)', 'red, blue, green'))

['red', 'blue', 'green']


In [10]:
import re
re.findall('((0\d{2}|\d{3})-(\d{8}|\d{6}))','复旦大学总机021-65642222')

[('021-65642222', '021', '65642222')]

In [13]:
import re
re.findall('(0\d{2}|\d{3})-(\d{8}|\d{6})','复旦大学总机021-65642222')

[('021', '65642222')]

* re模块的主要方法
  * 匹配和搜索
  * 分割字符串
  * 匹配和替换
  * 既可以直接使用re模块的方法进行字符串处理，也可以将模式编译为正则表达式对象，然后使用正则表达式的方法来操作字符串
* re模块的主要方法列表
  * pattern：匹配模式；string：要匹配的字符串；flag：匹配选项
  * findall(pattern,string\[,flags\])：列出字符串中模式的所有匹配项；返回匹配结果列表，若pattern含有组(子模式)，同时返回组的列表
  * search(pattern,string\[,flags\])：在字符串中寻找模式；若匹配，返回Match对象（参见4.2.5），否则返回None
  * match(pattern,string\[,flags\])：从字符串的开始处匹配模式；若匹配，返回Match对象，否则返回None
  * split(pattern,string\[,maxsplit=0\])：根据模式匹配项（匹配分割符）分割字符串，返回分割后的字符串列表，maxsplit为分割的最大次数
  * sub(pat,repl,string\[,count=0\])：将字符串中所有pat的匹配项用repl替换；并返回替换后的字符串，count为替换的最大次数
  * subn(pat,repl,string\[,count=0\])：将字符串中所有pat的匹配项用repl替换；并返回元组：(替换后的字符串,替换次数），count为替换的最大次数
  * escape(string)：将字符串中所有特殊正则表达式字符转义
* flag：匹配选项
  * re.l（re.IGNORECASE）：忽略大小写
  * re.L（re.LOCALE）：\w \W \b \B \s \S与本地字符集有关
  * re.M（re.MULTILINE）：多行匹配模式
  * re.S（re.DOTALL）：使元字符也匹配换行符
  * re.U（re.UNICODE）：匹配Unicode字符
  * re.X（re.VERBOSE）：忽略模式中的空格，并可以使用#注释，提高可读性

In [28]:
import re
text = 'alpha. beta....gamma. delta'
pat = '[a-zA-Z]+'
print(re.findall(pat,text)) #pat和text不能反过来
print(re.findall('done|quit','done'))

print(re.match('done|quit','doe!'))
print(re.match('to',"To be,\nor not to be")) # 只会从开头处开始匹配
# 除非
print(re.match('to', 'To be,\nor not to be', re.I)) # 忽略大小写

print(re.search('to',"To be,\nor not to be"))

print(re.split('[\. ]+',text))
print(re.split('[\. ]+', text, maxsplit=2))
print(re.split('[\. ]+', text, 1)) # 分割一次

['alpha', 'beta', 'gamma', 'delta']
['done']
None
None
<re.Match object; span=(0, 2), match='To'>
<re.Match object; span=(14, 16), match='to'>
['alpha', 'beta', 'gamma', 'delta']
['alpha', 'beta', 'gamma. delta']
['alpha', 'beta....gamma. delta']


* 示例：替换、转义

In [29]:
import re
pat = '{name}'
text = 'Dear {name}...'
print(re.sub(pat,'Mr.Dong',text)) #字符串替换

s = 'a s d'
print(re.sub('a|s|d','good',s)) #字符串替换

print(re.escape('http://www.python.org')) #字符串转义

Dear Mr.Dong...
good good good
http://www\.python\.org


* 示例：删除字符串中重复的空格

In [35]:
import re
s='aaa      bb      c d e   fff    '
print(re.split('[\s]+',s)) # \s匹配任何空白字符

print(re.split('[\s]+',s.strip())) # .strip()删除某端重复字符

print(' '.join(re.split('[\s]+',s.strip())))

print(re.sub('\s+',' ',s.strip()))

print(s)

print(s.split())     #也可以不使用正则表达式

print(' '.join(s.split()))

['aaa', 'bb', 'c', 'd', 'e', 'fff', '']
['aaa', 'bb', 'c', 'd', 'e', 'fff']
aaa bb c d e fff
aaa bb c d e fff
aaa      bb      c d e   fff    
['aaa', 'bb', 'c', 'd', 'e', 'fff']
aaa bb c d e fff


* 示例：使用以\开头的元字符

In [36]:
import re
example = 'ShanDong Institute of Business and Technology'
print(re.findall('\\ba.+?\\b',example)) # 以a开头的完整单词

print(re.findall('\\Bo.+?\\b',example)) # 不以o开头且含有o字母的单词剩余部分

print(re.findall('\\b\w.+?\\b',example)) # 所有单词

print(re.findall(r'\b\w.+?\b',example)) # 使用原始字符串，减少需要输入的符号数量

print(re.findall('\d\.\d\.\d','Python 2.7.8,Python 3.4.2')) # 查找并返回x.x.x的数字形式
 
print(re.split('\s',example))# 使用任何空白字符分割字符串

['and']
['ong', 'ology']
['ShanDong', 'Institute', 'of', 'Business', 'and', 'Technology']
['ShanDong', 'Institute', 'of', 'Business', 'and', 'Technology']
['2.7.8', '3.4.2']
['ShanDong', 'Institute', 'of', 'Business', 'and', 'Technology']


#### <b style="color: red;">4.2.4 使用正则表达式对象</b>
* 使用re模块的compile()方法将正则表达式编译生成正则表达式对象，然后再使用正则表达式对象提供的方法进行字符串处理。
  * regex=re.compile(pattern\[,flags\]) #编译生成正则表达式
* 使用编译后的正则表达式对象可以提高字符串处理速度
* regex.match(string\[, pos\[, endpos\]\])：在字符串开头或指定位置进行搜索，模式必须出现在字符串开头或指定位置；若匹配，返回Match对象，否则返回None
* regex.search(string\[, pos\[, endpos\]\])：在整个字符串中进行搜索；若匹配，返回Match对象，否则返回None
* regex.findall(string\[, pos\[, endpos\]\])：在字符串中查找所有符合正则表达式的字符串列表；返回匹配结果列表，若pattern含有组，同时返回组的列表
* regex.sub(repl,string\[,count=0\])：将字符串中所有pattern的匹配项用repl替换；并返回替换后的字符串，count为替换的最大次数
* regex.subn(repl,string\[,count=0\])：将字符串中所有pat的匹配项用repl替换，并返回元组：(替换后的字符串,替换次数），count为替换的最大次数
* regex.split(string\[, maxsplit = 0\])：根据模式匹配项（匹配分割符）分割字符串，返回分割后的字符串列表，maxsplit为分割的最大次数

In [37]:
import re
example = 'ShanDong Institute of Business and Technology'
pattern = re.compile(r'\bB\w+\b') # 以B开头的单词
print(pattern.findall(example))

pattern = re.compile(r'\w+g\b') # 以g结尾的单词
print(pattern.findall(example))

pattern = re.compile(r'\b[a-zA-Z]{3}\b') # 查找3个字母长的单词
print(pattern.findall(example))

print(pattern.match(example)) # match为从字符串开头开始匹配，所以不成功，没有返回值
print(pattern.search(example)) # search为在整个字符串中搜索，所以成功

pattern = re.compile(r'\b\w*a\w*\b') # 查找所有含有字母a的单词
print(pattern.findall(example))

['Business']
['ShanDong']
['and']
None
<re.Match object; span=(31, 34), match='and'>
['ShanDong', 'and']


* 示例：替换

In [42]:
import re
example = '''Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.'''

# 1.将以字母“b”和”B”开头的单词替换为“*”
pattern = re.compile(r'\bb\w*\b',re.I)
print(pattern.sub('*',example))
print()

# 2.将第一个以字母“b”和”B”开头的单词替换为“*”（只替换一次）
pattern = re.compile(r'\bb\w*\b',re.I)
print(pattern.sub('*',example,1))
print()

# 3.将第一个以字母“b”开头的单词替换为“*”
pattern = re.compile(r'\bb\w*\b')
print(pattern.sub('*',example,1))
print()

* is * than ugly.
Explicit is * than implicit.
Simple is * than complex.
Complex is * than complicated.
Flat is * than nested.
Sparse is * than dense.
Readability counts.

* is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.

Beautiful is * than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.



* 示例：分割字符串

In [43]:
example = r'one,two,three.four/five\six?seven[eight]nine|ten'
pattern = re.compile(r'[,./\\?[\]\|]') # 指定多个可能分割符
print(pattern.split(example))

example = r'one1two2three3four4file5six6seven7eight8nine9ten'
pattern = re.compile(r'\d+') # 使用数字作分割符
print(pattern.split(example))

example = r'one two    three  four,five.six.seven,eight,nine9ten'
pattern = re.compile(r'[\s,.\d]+') # 允许分隔符重复
print(pattern.split(example))

['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten']
['one', 'two', 'three', 'four', 'file', 'six', 'seven', 'eight', 'nine', 'ten']
['one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight', 'nine', 'ten']


#### 4.2.5 子模式与match对象
* 使用()表示一个子模式，即()内的内容作为一个整体出现
  * 例如’(red)+’可以匹配’redred’、’redredred‘等多个重复’red’的情况
* re.findall(pattern,string\[,flags\])
  * 返回匹配结果列表，若pattern含有组(子模式)，同时返回组的列表

In [44]:
telNumber = '''Suppose my Phone No. is 0535-1234567,
yours is 010-12345678, his is 025-87654321.'''
pattern = re.compile(r'(\d{3,4})-(\d{7,8})')
print(pattern.findall(telNumber))

print(re.findall(r'(\d{3,4})-(\d{7,8})',telNumber))

print(re.findall(r'((\d{3,4})-(\d{7,8}))',telNumber))

[('0535', '1234567'), ('010', '12345678'), ('025', '87654321')]
[('0535', '1234567'), ('010', '12345678'), ('025', '87654321')]
[('0535-1234567', '0535', '1234567'), ('010-12345678', '010', '12345678'), ('025-87654321', '025', '87654321')]


* search和match，返回的结果为匹配对象（Match object），使用Match对象的方法，可以进行匹配结果的处理，包括对子模式()匹配的处理

In [54]:
m=re.match(r'www\.(.*)\..{3}','www.python.orgggg')
print(m.group()) # 返回匹配的1个或多个子模式内容

print(m.group(0))

print(m.group(1))

print(m.start(1)) # 返回指定子模式内容的起始位置

print(m.end(1)) # 返回指定子模式内容的结束位置的前一个位置

print(m.span(1)) # 返回指定子模式内容的起始位置和结束位置的前一个位置

# group() 返回一个包含匹配的所子模式内容的元组
# groupdict() 返回包含匹配的所有命名子模式内容的字典

www.python.org
www.python.org
python
4
10
(4, 10)


* 示例：phone.py

In [65]:
import re
telNumber = '''Suppose my Phone No. is 0535-1234567,
yours is 010-12345678, his is 025-87654321.'''
pattern = re.compile(r'(\d{3,4})-(\d{7,8})')
index = 0
while True:
    matchResult = pattern.search(telNumber, index) # index为开始匹配的位置
    if not matchResult:
        break
    print ('-'*30)
    print ('Success:')
    for i in range(3):
        print ('Searched content:', matchResult.group(i),\
        ' Start from:',matchResult.start(i),'End at:',matchResult.end(i),\
          ' Its span is:', matchResult.span(i)) # 记住.group(0)到.group(2)分别代表什么
    index = matchResult.end(2) # .end()为结束位置的l

------------------------------
Success:
Searched content: 0535-1234567  Start from: 24 End at: 36  Its span is: (24, 36)
Searched content: 0535  Start from: 24 End at: 28  Its span is: (24, 28)
Searched content: 1234567  Start from: 29 End at: 36  Its span is: (29, 36)
------------------------------
Success:
Searched content: 010-12345678  Start from: 47 End at: 59  Its span is: (47, 59)
Searched content: 010  Start from: 47 End at: 50  Its span is: (47, 50)
Searched content: 12345678  Start from: 51 End at: 59  Its span is: (51, 59)
------------------------------
Success:
Searched content: 025-87654321  Start from: 68 End at: 80  Its span is: (68, 80)
Searched content: 025  Start from: 68 End at: 71  Its span is: (68, 71)
Searched content: 87654321  Start from: 72 End at: 80  Its span is: (72, 80)


In [70]:
import re
telNumber = '''1231-1234567.'''
pattern = re.compile(r'(\d{3,4})-(\d{7,8})')
index = 0

a = pattern.search(telNumber, index)

print(a.end(2))

12
